<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Assign Subclass and Class Names to Cells in Single-Cell RNA-Seq Data — subclass_naming • CSSG.toolkit</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Assign Subclass and Class Names to Cells in Single-Cell RNA-Seq Data — subclass_naming"><meta property="og:description" content="This function assigns meaningful names to subclasses and classes based on clustering and heterogeneity markers in a single-cell RNA sequencing project. It uses information about heterogeneity, class markers, and subclass markers to perform the annotation. It updates the subclass names in the `scRNAproject` object by combining cluster names with subclass names, based on the provided markers and species. This function assumes that clustering information and heterogeneity markers are already present in the `scRNAproject` metadata.
The structure of manually curated marker data, including marker class and marker subclass, is available on the GitHub page: https://github.com/jkubis96/CSSG. If set to NULL by default, the names will be based on specific markers for the clusters."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">CSSG.toolkit</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"></ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Assign Subclass and Class Names to Cells in Single-Cell RNA-Seq Data</h1>

    <div class="hidden name"><code>subclass_naming.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This function assigns meaningful names to subclasses and classes based on clustering and heterogeneity markers in a single-cell RNA sequencing project. It uses information about heterogeneity, class markers, and subclass markers to perform the annotation. It updates the subclass names in the `scRNAproject` object by combining cluster names with subclass names, based on the provided markers and species. This function assumes that clustering information and heterogeneity markers are already present in the `scRNAproject` metadata.
The structure of manually curated marker data, including marker class and marker subclass, is available on the GitHub page: https://github.com/jkubis96/CSSG. If set to NULL by default, the names will be based on specific markers for the clusters.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">subclass_naming</span><span class="op">(</span></span>
<span>  <span class="va">sc_project</span>,</span>
<span>  class_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  subclass_markers <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  species <span class="op">=</span> <span class="st">"Homo sapiens"</span>,</span>
<span>  chunk_size <span class="op">=</span> <span class="fl">5000</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>


<dl><dt id="arg-sc-project">sc_project<a class="anchor" aria-label="anchor" href="#arg-sc-project"></a></dt>
<dd><p>An object of class `scRNAproject`, which contains the data and metadata required for subclass and class naming. It should include matrices with normalized gene expression data and metadata related to heterogeneity markers.</p></dd>


<dt id="arg-species">species<a class="anchor" aria-label="anchor" href="#arg-species"></a></dt>
<dd><p>A character string specifying the species (default: 'Homo sapiens'), which determines the structure of the names.</p></dd>


<dt id="arg-chunk-size">chunk_size<a class="anchor" aria-label="anchor" href="#arg-chunk-size"></a></dt>
<dd><p>The number of cells to process in each chunk during aggregation (default is 5000).</p></dd>


<dt id="arg-markers-class">markers_class<a class="anchor" aria-label="anchor" href="#arg-markers-class"></a></dt>
<dd><p>A list of class markers (default: NULL). If NULL the cluster will obtain native names from 'sc_project@names$primary' or 'sc_project@names$cluster'</p></dd>


<dt id="arg-markers-subclass">markers_subclass<a class="anchor" aria-label="anchor" href="#arg-markers-subclass"></a></dt>
<dd><p>A list of subclass markers (default: NULL). If NULL the markers will used from 'sc_project@metadata$heterogeneity_markers'</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>An updated `scRNAproject` object with new subclass names assigned. The subclass names will be stored in the `names$subclass` slot of the `sc_project`.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>The function first checks if heterogeneity markers are present in the metadata. If markers are missing or invalid, it raises an error. The subclass names are then assigned by:
- Using the class markers to determine class names for the clusters.
- Using the subclass markers and heterogeneity markers to assign meaningful subclass names to cells.
- Combining the class and subclass names to form the final names for each cell.</p>
<p>The `aggregation_chr` and `aggregation_num` functions are used to handle different types of data (character vs numeric) during subclass aggregation.</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Assuming you have a valid scRNAproject object (`sc_project`), class markers, and subclass markers</span></span></span>
<span class="r-in"><span><span class="va">updated_project</span> <span class="op">&lt;-</span> <span class="fu">subclass_naming</span><span class="op">(</span><span class="va">sc_project</span>, class_markers <span class="op">=</span> <span class="cn">NULL</span>, subclass_markers <span class="op">=</span> <span class="cn">NULL</span>, species <span class="op">=</span> <span class="st">'Homo sapiens'</span>, chunk_size <span class="op">=</span> <span class="fl">5000</span><span class="op">)</span></span></span>
<span class="r-err co"><span class="r-pr">#&gt;</span> <span class="error">Error:</span> object 'sc_project' not found</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Jakub Kubiś - JBioSystem.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

      </footer></div>






  </body></html>

